Improve Scopes doc, put under Explanation

nojb opened the pull request:
Following #11829, I tried to give a bit more substance to the "Scopes" doc, and moved it to the "Getting Started" section, since it is pretty important information to make sense of how Dune operates.

Fixes #11829

nojb committed:
Improve Scopes doc, put under Getting Started

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

gasche commented on doc/getting-started/scopes.rst:
```diff
@@ -0,0 +1,32 @@
+or by the first :doc:`/reference/dune-workspace/index` file found when climbing
+up the directory hierachy starting from the current directory (see
+:ref:`finding-root`). It can also be specified explicitly by passing the
+``--root`` flag (see :ref:`forcing-root`).
```
I found this paragraph a bit unclear (you say that the root is defined by A or B, but it's not clear which applies if both make sense; and I'm not sure that I understand the descriptions of A and B). In contrast, https://dune.readthedocs.io/en/stable/usage.html#finding-the-root is clear. I would less here and redirect to that part instead. (It's not clear to me why "Finding the root" is in the "Command-line Interface" page, but oh well.)

gasche commented on doc/getting-started/scopes.rst:
```diff
@@ -0,0 +1,32 @@
+Scopes
```
Is "Scopes" the right title for this page? I'm not sure what the word refers to. Maybe if you tried to summarize what the page is about you would find a more descriptive title?

gasche reviewed (commented):
I believe this is an improvement, thanks!

- You mention `dune-workspace` and `dune-project` files in scopes.rst, but not `dune` files. It might be worth mentioning them briefly.
- I think it would be useful to have backlinks from the `dune`, `dune-project` and `dune-workspace` reference pages back to this page. (I'm not sure whether you have done that from the markup alone.) When I wanted to clarify the nested semantics of `dune-project` file, I went to look at the `dune-project` reference file.

nojb committed:
Backlink from dune/dune-project/dune-workspace

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

nojb committed:
Mention dune files

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

nojb committed:
Rename Scopes => Dune projects and workspaces

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

nojb committed:
Tweak

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

nojb committed:
Fix

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

nojb commented:
Thanks for the feedback @gasche!

> * You mention `dune-workspace` and `dune-project` files in scopes.rst, but not `dune` files. It might be worth mentioning them briefly.

Fixed: I added a brief blurb about them.

> * I think it would be useful to have backlinks from the `dune`, `dune-project` and `dune-workspace` reference pages back to this page. (I'm not sure whether you have done that from the markup alone.) When I wanted to clarify the nested semantics of `dune-project` file, I went to look at the `dune-project` reference file.

Fixed: I added links in each of these sections back to the one containing the new information.

nojb committed:
Continue

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

nojb committed:
Reduce diff

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

nojb committed:
Fix

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

gasche commented on doc/scopes.rst:
```diff
@@ -0,0 +1,30 @@
+directory or by the presence of a :doc:`/reference/dune-workspace/index` file
+(see :ref:`finding-root` and :ref:`forcing-root` for details). All
+:doc:`/reference/dune-workspace/index` files other than the one specifying the
+root of the workspace are ignored.
```
The last sentence is about `dune-workspace`, it could go to the dedicated paragraph currently on lines 24-25.

nojb replied:
Fixed, thanks.

gasche commented on doc/scopes.rst:
```diff
@@ -0,0 +1,30 @@
+inherited by all Dune projects in the workspace.
+
+Within a Dune project, :doc:`/reference/dune/index` files are used to define
+all objects of interest for Dune: libraries, executables, tests, etc. Only
```
You could mention that `dune` files are relative to their current directory, and that there typically are many `dune` files in a single project.

nojb replied:
Good idea. Fixed. Thanks.

Alizter replied:
It is also possible to use the `(subdir)` stanza. https://dune.readthedocs.io/en/stable/reference/dune/subdir.html

Ultimately each stanza does end up living in its own directory, but there is no such strict for the dune file.

nojb replied:
Fair enough, I have added a comment to this effect (I prefer not to explain the notion that it is the stanza that is attached to a directory, rather than the Dune file because it is a bit subtle and only relevant for advanced uses).

gasche reviewed (approved):
This is fine as far as I'm concerned -- see minor comments. It does solve the issue I created. Thanks!

nojb committed:
Continue

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

gasche commented on doc/scopes.rst:
```diff
@@ -0,0 +1,35 @@
+Within a Dune project, :doc:`/reference/dune/index` files are used to define all
+objects of interest for Dune: libraries, executables, tests, etc. All references
+within :doc:`/reference/dune/index` files are always written relative to the
+directory containing the file.  There are typically many Dune files in a Dune
```
```suggestion
directory containing the file.  There are typically many :doc:`/reference/dune/index` files in a Dune
```

nojb replied:
Thanks, fixed.

nojb committed:
Mention (subdir)

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

nojb committed:
Fix link

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

maiste commented on doc/scopes.rst:
Wouldn't it be better if this put this under the `Explanation` section?

nojb replied:
My view is that this is pretty crucial information to understand Dune's behaviour, so under "Getting Started" seems OK. But don't have a very strong opinion on the subject.

Leonidas-from-XIV replied:
I concur with @maiste on this one. When you're getting started you usually have one `dune-project` and no `dune-workspace` so I don't think that the understanding of the mechanism is particularly important for getting started.

maiste reviewed (commented):
Thanks for the PR! I add a quick look. I'll try to have a closer one later.

nojb committed:
Merge remote-tracking branch 'upstream/main' into scopes_doc

nojb committed:
Move to Explanations

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

nojb committed:
Merge branch 'main' into scopes_doc

nojb commented:
Friendly ping @maiste. Are you happy with the current state of the PR?

nojb committed:
Move scopes to explanation/

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

nojb commented on doc/explanation/index.rst:
```diff
@@ -7,6 +7,7 @@ the rest of the OCaml ecosystem.
 .. toctree::
    :maxdepth: 1
 
+   ../scopes
```
OK, fixed, thanks.

nojb committed:
Fix

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

Alizter commented on doc/explanation/scopes.rst:
```diff
@@ -0,0 +1,36 @@
+
+A Dune project is defined by the presence of a
+:doc:`/reference/dune-project/index` file. Each Dune project extends over the
+subtree rooted at the directory containing the
```
"subtree" sounds like we are in a subdirectory when instead we are really talking about the directory at which the dune-project exists. I would probably reword that to "file tree".

nojb replied:
Amended, thanks.

Alizter commented on doc/explanation/scopes.rst:
```diff
@@ -0,0 +1,36 @@
+Dune projects and workspaces
```
```suggestion
Dune Projects and Workspaces
```

Alizter replied:
We use title case everywhere expect with headers that are about dune files like `dune` and `dune-project`.

Alizter commented on doc/explanation/scopes.rst:
```diff
@@ -0,0 +1,36 @@
+subtree rooted at the directory containing the
+:doc:`/reference/dune-project/index` file, excluding any nested Dune projects.
+
+The root of the current workspace is determined by the outermost
+:doc:`/reference/dune-project/index` file in an ancestor of the current
+directory or by the presence of a :doc:`/reference/dune-workspace/index` file
+(see :ref:`finding-root` and :ref:`forcing-root` for details).
```
```
Dune determines the root of the current workspace by finding the topmost ancestor directory containing a `dune-project` file or by the presence of a `dune-workspace` file.
```

An alternative wording, feel free to pick and choose. Q: Do we always pick the topmost dune-workspace file? What happens if we have A/B where A contains a dune-project and B a dune-workspace?

Alizter commented on doc/explanation/scopes.rst:
```diff
@@ -0,0 +1,36 @@
+directory or by the presence of a :doc:`/reference/dune-workspace/index` file
+(see :ref:`finding-root` and :ref:`forcing-root` for details).
+
+Different Dune projects within the same Dune workspace are independent of each
+other and no settings are shared between them, even if they are nested within
+each other.
```
What about dune language versions? I believe we take the maximum of all the dune-lang versions. We should also check if the same holds for syntax extensions.

If my dune-project has dune lang 3.18 can I vendor a project with dune lang 3.19 in its dune-project file? Probably not.

Alizter commented on doc/explanation/scopes.rst:
```diff
@@ -0,0 +1,36 @@
+other and no settings are shared between them, even if they are nested within
+each other.
+
+Settings in :doc:`/reference/dune-workspace/index`, on the other hand, are
+inherited by all Dune projects in the workspace. Note that all
+:doc:`/reference/dune-workspace/index` files other than the one specifying the
+root of the workspace are ignored.
```
Maybe worth mentioning that dune-workspace files have different settings and configuration options than dune-project although there is some overlap. There is also dune config which I believe overrides the dune-workspace file but I am not sure.

Alizter commented on doc/explanation/scopes.rst:
```diff
@@ -0,0 +1,36 @@
+Within a Dune project, :doc:`/reference/dune/index` files are used to define all
+objects of interest for Dune: libraries, executables, tests, etc. References
+within :doc:`/reference/dune/index` files are resolved relative to the directory
+containing the file (except if using a :doc:`/reference/dune/subdir` stanza).
```
Or alternatively `chdir` inside a user action, but this might not be worth mentioning.

Alizter reviewed (commented):
Thanks @nojb for starting this. I have a few review comments and I will do another pass after those are done.

I would like for this document to be maximally informative to those who come across it and not have them with more questions than when they started.

nojb committed:
Update doc/explanation/scopes.rst

Co-authored-by: Ali Caglayan <alizter@gmail.com>
Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

nojb committed:
subtree => file tree

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

nojb committed:
Wording

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

nojb committed:
Mention overlap of dune-workspace and dune-project

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

nojb committed:
Merge branch 'main' into scopes_doc

nojb commented:
Thanks for the careful review @Alizter! I think I addressed all points. Let me know if you have any additional feedback.

Alizter commented on doc/explanation/scopes.rst:
```diff
@@ -0,0 +1,39 @@
+Dune workspace is a set of Dune projects. A typical workspace consists of a
+single Dune project.
+
+A Dune project is defined by the presence of a
```
Just as we used italics for Dune workspace, we should use italics for Dune project since we are defining the term.

nojb replied:
Fixed, thanks.

nojb committed:
Use italics

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

maiste commented on doc/explanation/scopes.rst:
```diff
@@ -0,0 +1,39 @@
+within :doc:`/reference/dune/index` files are resolved relative to the directory
+containing the file (except if using a :doc:`/reference/dune/subdir` stanza or
+the :doc:`/reference/actions/chdir` action`). There are typically many
+:doc:`/reference/dune/index` files in a Dune project (one per directory, unless
+the directory does not contain anything relevant to Dune).
```
```suggestion
:doc:`/reference/dune/index` files in a Dune project (one per directory, unless
the directory does not contain anything relevant to Dune or the sub directory
is a descendant of a directory using :doc:`/reference/dune/include_subdirs`).
```
Should we add a reference to `(include_subdirs [...])`? I'm asking because this is a useful and underrated feature and it has its place in this part of the documentation. (I'm not sure if the suggestion is correctly working...)

nojb replied:
Actually, I would rather keep in this section only the explanations of the general case, and indicate that these general rules have exceptoins, pointing the reader to the specific stanzas (`(subdir)`, `(include_subdirs)`, `(chdir)`, etc) to learn about the details.

I will push a commit in this direction shortly.

nojb replied:
Addressed in 6602c2a02abf2737f51e08f031e9b167ca6c827b

nojb committed:
Continue

Signed-off-by: Nicolás Ojeda Bär <n.oje.bar@gmail.com>

maiste reviewed (commented):
I have done another round with minor questions and suggestions. Thanks for taking care of this! :smile:

nojb commented:
Thanks for the feedback @maiste. However, both of your suggestions concern other pages of the documentation that the one being addressed in this PR (only a backlink was added in this PR). I would rather that a dedicated PR be opened for them, maybe with further changes related to package management or any other needed information. Is that OK?

maiste reviewed (approved):
Oh I missed that, sorry! Sure no problem 😊
Under those conditions, LGTM!

nojb committed:
Merge branch 'main' into scopes_doc

Alizter reviewed (approved):
LGTM, merge when you wish.

nojb commented:
Thank you everyone for the feedback!
